util/dvsim/utils.py

```diff
@@ -86,84 +95,280 @@ def parse_hjson(hjson_file):
+        return _subst_wildcards(var, mdict, ignored_wildcards, ignore_error, [])[0]
+    except ValueError as err:
+        log.error(str(err))
+        sys.exit(1)
+
+
+def test_subst_wildcards():
+    '''Pytest-compatible test for the subst_wildcards function.'''
+    # Basic checks
+    assert subst_wildcards('foo {x} baz', {'x': 'bar'}) == 'foo bar baz'
+
+    # Stringify
+    assert (subst_wildcards('{a}, {b}, {c}, {d}',
+                            {'a': 'a', 'b': True, 'c': 42, 'd': ['{b}', 10]}) ==
+            'a, 1, 42, 1 10')
+
+    # Ignored wildcards (with or without a match in mdict)
+    assert (subst_wildcards('{a} {b}', {'a': 'aye', 'b': 'bee'},
+                            ignored_wildcards=['a']) ==
+            '{a} bee')
+    assert (subst_wildcards('{a} {b}', {'b': 'bee'},
+                            ignored_wildcards=['a']) ==
+            '{a} bee')
+
+    # Environment variables. We will always have PWD and can probably assume
+    # that this won't itself have any braced substrings.
+    assert (subst_wildcards('{PWD}', {}) == os.environ['PWD'])
+
+    # Missing variable with ignore_error=False, running _subst_wildcards
+    # instead so that we can catch the error. (We assume that 'biggles' isn't
+    # in the environment)
+    with pytest.raises(ValueError) as excinfo:
+        _subst_wildcards('{biggles} {b}', {'b': 'bee'}, [], False, [])
+    assert "unknown wildcard, '{biggles}'" in str(excinfo.value)
+
+    # ignore_error=True.
+    assert (subst_wildcards('{biggles} {b}', {'b': 'bee'},
+                            ignore_error=True) ==
+            '{biggles} bee')
+
+    # Check we support (non-circular) recursion
+    assert (subst_wildcards('{a}', {'a': '{b}', 'b': 'c'}) == 'c')
+
+    # Check we spot circular recursion
+    with pytest.raises(ValueError) as excinfo:
+        _subst_wildcards('{a}', {'a': '{b}', 'b': '{a}'}, [], False, [])
+    assert "circular expansion of wildcard '{a}'" in str(excinfo.value)
+
+    # Circular recursion with ignore_error
+    assert (subst_wildcards('{a}',
+                            {'a': 'A{b}', 'b': 'B{a}'}, ignore_error=True) ==
+            'AB{a}')
+
+    # Computed variable names (probably not a great idea, but it's probably
+    # good to check this works the way we think)
+    assert subst_wildcards('{a}b}', {'a': 'a {', 'b': 'bee'}) == 'a bee'
+
+    # Some eval_cmd calls (using echo, which should always work)
+    assert (subst_wildcards('{eval_cmd}echo foo {b}', {'b': 'bar'}) ==
+            'foo bar')
+
+    # Make sure that nested commands work
+    assert (subst_wildcards('{eval_cmd} {eval_cmd} echo echo a', {}) == 'a')
+
+    # Recursive expansion
+    assert (subst_wildcards('{var}',
+                            {
+                                'var': '{{foo}_xyz_{bar}}',
+                                'foo': 'p',
+                                'bar': 'q',
+                                'p_xyz_q': 'baz'
+                            }) == 'baz')
```

Comment:
Consider adding tests for `find_and_substitute_wildcards()` as well; example: `sub_dict == full_dict == '{ a: [ { b: "{b}" } ], c: "{c}" }` should detect loops both on b and c (I suspect the current code will throw value-not-found error instead, which will be ignored if `ignore_error` is set). 

Would you also please move the tests to a separate python source?

Reply:
I just checked and, without `ignore_error` this complains about a missing definition of `b`. With `ignore_error` (and changing the code to always fail on a circular reference), this fails because of the circular reference for `c`. You're right that we're expanding the `b` part in an odd way, but that's a bug/weird spec of `find_and_substitute_wildcards`, which isn't the function this PR was trying to tidy up.

Since this PR is all about `subst_wildcards`, I don't really want to add a testsuite for (and fix bugs in) `find_and_substitude_wildcards` as part of it. I'm totally happy to do that as a follow-up.

I'll happily move the code to a separate file. Would you prefer `util/dvsim/tests/utils_test.py`, `util/dvsim/utils_test.py` or something else?

Reply:
`util/dvsim/utils_test.py` sgtm.

Reply:
Cool. I've pulled that out and put a (completely trivial) Makefile with a "check" target in the dvsim directory, mainly to remind us how to run pytest :-)